pkg/ring/ring.go

```diff
@@ -335,20 +338,49 @@ func (r *Ring) GetAll(op Operation) (ReplicationSet, error) {
 	maxUnavailable := r.cfg.ReplicationFactor / 2
 	numRequired -= maxUnavailable
 
-	ingesters := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	maxUnavailableZones := len(r.ringZones) - maxUnavailable - 1
+	if maxUnavailableZones < 0 {
+		maxUnavailableZones = 0
+	}
+
+	instances := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	zoneFailures := make(map[string]int)
 	for _, ingester := range r.ringDesc.Ingesters {
 		if r.IsHealthy(&ingester, op) {
-			ingesters = append(ingesters, ingester)
+			instances = append(instances, ingester)
+		} else {
+			zoneFailures[ingester.Zone]++
+		}
+	}
+
+	if r.cfg.ZoneAwarenessEnabled && len(zoneFailures) > 0 {
+		if len(zoneFailures) > maxUnavailableZones {
+			return ReplicationSet{}, ErrTooManyFailedIngesters
+		}
+
+		filteredInstances := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
```

Comment:
I would add a comment explaining why we do this. Something like this: "Given we can tolerate a zone failure, we can entirely skip querying instances in that zone."

Reply:
The comment was added.